feat:(kms) encrypt data with DEK using AES-GCM instead of AES-CBC

aramase opened the pull request:
Signed-off-by: Anish Ramasekar <anish.ramasekar@gmail.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
Follow-up to https://github.com/kubernetes/kubernetes/pull/108745. This change updates the KMS envelope encryption to be use AES-GCM to encrypt data using the DEK instead of AES-CBC. To allow for downgrades and HA upgrades, this functionality allows reads with AES-GCM and AES-CBC.

- xref https://github.com/kubernetes/kubernetes/issues/81127
- This PR supersedes https://github.com/kubernetes/kubernetes/pull/110011/

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Encrypt data with DEK using AES-GCM instead of AES-CBC for kms data encryption. No user action required. Reads with AES-GCM and AES-CBC will continue to be allowed.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
https://kubernetes.io/docs/tasks/administer-cluster/kms-provider/
```

aramase committed:
feat:(kms) encrypt data with DEK using AES-GCM instead of AES-CBC

Signed-off-by: Anish Ramasekar <anish.ramasekar@gmail.com>

aramase commented:
/sig auth
/hold

aramase commented:
/cc @enj @ritazh

sftim commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -369,8 +369,8 @@ func secretboxPrefixTransformer(config *apiserverconfig.SecretboxConfiguration)
 		// v1.24: write using AES-CBC only but support reads via AES-CBC and AES-GCM (so we can move to AES-GCM)
-		// TODO(aramase): swap this ordering in v1.25
-		return unionTransformers{aestransformer.NewCBCTransformer(block), aestransformer.NewGCMTransformer(block)}
+		// v1.25: write using AES-GCM only but support reads via AES-GCM and fallback to AES-CBC for backwards compatibility
```
Why does this comment mention a specific Kubernetes release? Should we keep this comment as-is for v1.26 and later?

sftim reviewed (commented):
I'm wondering how we'll document this.

The text to change in the docs is something like:
> Uses envelope encryption scheme: Data is encrypted by data encryption keys (DEKs) that use AES-CBC with [PKCS#7](https://datatracker.ietf.org/doc/html/rfc2315) padding; DEKs are encrypted by key encryption keys (KEKs) according to configuration in Key Management Service (KMS)

That text is already kind of complicated. It'll become even more complicated with the change in this PR.

- Should this revised encryption be behind a feature gate? (IMO: yes)
- When will we drop support for reading the previous envelope encryption scheme
   - or will we support readback via AES-CBC indefinitely 
- What tasks should a cluster operator perform to get ready for that removal (if planned)

I think I'd prefer to let the cluster operator _specify_ a DEK encryption mechanism for writes and configure which encryption mechanisms are supported for readback. That's similar to how a cluster operator can configure eg `aesgcm` for writes and `aesgcm`, `kms` or `aescbc` for reads.

sftim commented:
The “Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:” is blank. This is a significant change for some cluster operators and we should include a link to relevant docs. Ideally there is also a KEP that we should link to.

Having a KEP also triggers docs tracking; without it, SIG Release won't know to nag folks about getting the docs updated.

sftim commented:
@aramase this is held - can you add a comment to clarify why and under what circumstances you'd expect to unhold?

leilajal commented:
/triage accepted

mikedanese commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -369,8 +369,8 @@ func secretboxPrefixTransformer(config *apiserverconfig.SecretboxConfiguration)
 func envelopePrefixTransformer(config *apiserverconfig.KMSConfiguration, envelopeService envelope.Service, prefix string) (value.PrefixTransformer, error) {
 	baseTransformerFunc := func(block cipher.Block) value.Transformer {
 		// v1.24: write using AES-CBC only but support reads via AES-CBC and AES-GCM (so we can move to AES-GCM)
-		// TODO(aramase): swap this ordering in v1.25
```
Keep a TODO to complete the migration?

aramase replied:
Added a TODO to drop AES CBC in v1.26

ritazh commented:
perhaps update release note:
> Encrypt data with DEK using AES-GCM instead of AES-CBC for kms data encryption. Not user action required. Reads with AES-GCM and AES-CBC will continue to be allowed.

mikedanese commented:
From sig-auth: We consider this to be an implementation detail of the envelope encryption storage transformer. There are no feature gates or user knobs and the user should never be aware of the migration. Let's make sure the doc is accurate but I don't think we need to stretch this out with a KEP / features gates / configuration at this time.

aramase commented:
/priority important-soon

aramase commented:
- I've modified some of the tests we added in the previous release to simulate downgrade (write with AES CBC). @enj LMK what you think.
- Updated the release note

/hold cancel

aramase commented:
/milestone v1.25

sftim commented:
For the release note
> Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

then could we link to https://kubernetes.io/docs/tasks/administer-cluster/kms-provider/? That saves readers a job.

sftim commented:
We might want to document how to ensure all relevant objects use an AES-GCM DEK (which I think is: read them all, and then rewrite them - same as if you migrate to / or from KMS encryption at rest)

aramase commented:
@enj @mikedanese This PR is ready for review. Could you take a look when you get a chance?

aramase commented:
@mikedanese @enj Could we get this PR reviewed soon so we don't miss the v1.25 code freeze?

aramase commented:
The v1.25 code freeze is tomorrow.

@mikedanese Could you review this so we can get it in for v1.25?

jeremyrickard commented:
/milestone v1.25

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -369,8 +369,9 @@ func secretboxPrefixTransformer(config *apiserverconfig.SecretboxConfiguration)
-		// TODO(aramase): swap this ordering in v1.25
-		return unionTransformers{aestransformer.NewCBCTransformer(block), aestransformer.NewGCMTransformer(block)}
+		// v1.25: write using AES-GCM only but support reads via AES-GCM and fallback to AES-CBC for backwards compatibility
+		// TODO(aramase): v1.26+: drop support for AES-CBC
```
We cannot drop CBC read support until we automate storage migration.

We could have a release note that hard require that storage migration be performed, but seems a bit user hostile to me.

aramase replied:
Noted! Added this as a comment.

enj reviewed (changes requested):
The code change itself LGTM.

enj commented:
/lgtm
/approve
